components/places/src/storage/history.rs

```diff
@@ -1144,6 +1144,35 @@ pub fn get_visit_page(
     Ok(HistoryVisitInfos { infos })
 }
 
+pub fn get_visit_page_with_bound(
+    db: &PlacesDb,
+    offset: i64,
+    bound: i64,
+    count: i64,
+    exclude_types: VisitTransitionSet,
+) -> Result<HistoryVisitInfos> {
+    let allowed_types = exclude_types.complement();
+    let infos = db.query_rows_and_then_named_cached(
+        "SELECT h.url, h.title, v.visit_date, v.visit_type, h.hidden
+         FROM moz_places h
+         JOIN moz_historyvisits v
+           ON h.id = v.place_id
+         WHERE ((1 << v.visit_type) & :allowed_types) != 0 AND
+               NOT h.hidden AND v.visit_date <= :bound
+         ORDER BY v.visit_date DESC, v.id
+         LIMIT :count
+         OFFSET :offset",
+        rusqlite::named_params! {
+            ":allowed_types": allowed_types,
+            ":bound": bound,
+            ":count": count,
+            ":offset": offset,
+        },
+        HistoryVisitInfo::from_row,
+    )?;
+    Ok(HistoryVisitInfos { infos })
+}
+
```

Comment:
Please also consider adding some tests for the new `get_visit_page_with_bound` function, copying the style of the tests further down in this file.  Interesting edge-cases to consider include:

* Does it work correctly when `count` is very small, and there are more than `count` items with the same `visit_date`?
* Does it work correctly when `offset` happens to fall precisely on a point where the `visited_date` changes? When it falls one item on either side?

Reply:
Create some tests under `test_get_visit_page_with_bound`.